binding/rust/src/porcupine.rs

```diff
@@ -0,0 +1,383 @@
+                    return Err(PorcupineError::new(
+                        PorcupineErrorStatus::ArgumentError,
+                        &format!(
+                            "Couldn't find keyword file at {}",
+                            keyword_path.as_ref().display()
+                        ),
+                    ));
+                }
+            }
+
+            for sensitivity in sensitivities {
+                if !(0.0 <= *sensitivity && *sensitivity <= 1.0) {
+                    return Err(PorcupineError::new(
+                        PorcupineErrorStatus::ArgumentError,
+                        "A sensitivity value should be within [0, 1]",
+                    ));
+                }
+            }
+
+            let lib = Library::new(library_path.as_ref())
+                .expect("Failed to load porcupine dynamic library");
+
+            let pv_porcupine_init: PvPorcupineInitFn = lib
+                .get(b"pv_porcupine_init")
+                .expect("Failed to load init function symbol from Porcupine library");
+
+            let pv_model_path = pathbuf_to_cstring(&model_path);
+            let pv_keyword_paths = keyword_paths
+                .iter()
+                .map(|keyword_path| pathbuf_to_cstring(keyword_path))
+                .collect::<Vec<_>>();
+            let pv_keyword_paths_ptrs = pv_keyword_paths
+                .iter()
+                .map(|keyword_path| keyword_path.as_ptr())
+                .collect::<Vec<_>>();
+            let mut cporcupine = std::ptr::null_mut();
+
+            let status = pv_porcupine_init(
+                pv_model_path.as_ptr(),
+                pv_keyword_paths.len() as i32,
+                pv_keyword_paths_ptrs.as_ptr(),
+                sensitivities.as_ptr(),
+                addr_of_mut!(cporcupine),
+            );
+            if status != PicovoiceStatuses::SUCCESS {
+                panic!(
+                    "{}",
+                    PorcupineError::new(
+                        PorcupineErrorStatus::LibraryError(status),
+                        "Failed to initialize the Porcupine library",
+                    )
+                );
+            }
+
+            let pv_sample_rate: PvSampleRateFn = lib
+                .get(b"pv_sample_rate")
+                .expect("Failed to load function symbol from Porcupine library");
+
+            let pv_porcupine_frame_length: PvPorcupineFrameLengthFn = lib
+                .get(b"pv_porcupine_frame_length")
+                .expect("Failed to load function symbol from Porcupine library");
+
+            let sample_rate = pv_sample_rate();
+            let frame_length = pv_porcupine_frame_length();
+
+            return Ok(Self {
+                cporcupine,
+                lib,
+                sample_rate,
+                frame_length,
+            });
+        }
+    }
+
+    pub fn process(&self, pcm: &[i16]) -> Result<i32, PorcupineError> {
+        if pcm.len() as i32 != self.frame_length {
```

Comment:
Nope :D 
Rust won't let you do that!

Reply:
In this case the `&[..]` syntax is called a slice, which doesn't represent a type per-say but a "view" into a supported type (Like a vector). Since Rust operates entirely on moves/references, the "borrow-checker" can guarantee at compile-time that this view will never be `null` at runtime.